Data renderer

eadf opened the pull request:
As described in #12

The normal image renderer is now feature gated, but enabled by default. 

The data renderer is also behind a feature gate, but it does not have any dependencies and does only consists of a few bytes of code. Maybe that feature gate should be dropped (remove every line with `#[cfg(feature = "data_renderer")]` )

Edit: Look what your code did to blender :) 
![metavolume](https://raw.githubusercontent.com/eadf/toxicblend.rs/lsystem/img/lsystems.png)
![metavolume](https://raw.githubusercontent.com/eadf/toxicblend.rs/lsystem/img/plant.png)

eadf committed:
bump dep. versions

eadf committed:
Added a feature gated data-only renderer.
Feature gated the default image-renderer

eadf commented:
I'm having problems propagating feature flags from main level `dcc-lsystem` down to the workspace `dcc-lsystem/dcc-lsystem` where they have effect.

Maybe hold this PR until we figure out a solution.

Edit: False alarm. I was trying to import from git, but cargo could not find `dcc-lsystem/dcc-lsystem` under the git path. 
```
[dependencies]
...
dcc-lsystem = { git = "https://github.com/eadf/dcc-lsystem", default-features=false, features=["data_renderer"]}
```
It should be ok when `dcc-lsystem/dcc-lsystem` is published as a non-virtual crate.

dcchut commented on dcc-lsystem/src/renderer.rs:
```diff
@@ -465,3 +489,34 @@ impl<Q: TurtleContainer> Renderer<ImageRendererOptions> for TurtleRenderer<Q> {
     }
 }
+
+#[cfg(feature = "data_renderer")]
```
As you said, this isn't pulling in any extra deps or adding a whole bunch of code so I think we can drop the `data_renderer` flag.

eadf replied:
Ok, fixed.

dcchut commented on dcc-lsystem/src/renderer.rs:
```diff
@@ -1,17 +1,27 @@
 use std::collections::HashMap;
 
+#[cfg(feature = "image_renderer")]
```
I haven't fully thought it through, but it might be work making a new module containing all of the image_renderer specific code.  I _think_ backwards compat can be maintained by doing a re-export?  e.g.:
```rust
// in `lib.rs`
#[cfg(feature = "image_renderer")]
pub mod whatever;

// in `renderer.rs:
#[cfg(feature = "image_renderer")]
pub use crate::whatever::ImageRendererOptions
// etc
```

eadf replied:
Yes, that works without needing to change any of the examples. Good idea.

It looks so much better without all those `#[cfg(feature = "image_renderer")]` everywhere.
Had to add two `pub (crate)` clauses on two (previously private) renderer fields.

dcchut commented on dcc-lsystem/src/renderer.rs:
```diff
@@ -465,3 +489,34 @@ impl<Q: TurtleContainer> Renderer<ImageRendererOptions> for TurtleRenderer<Q> {
+        self.compute(system.get_state());
+
+        // TODO: find a way to move lines() instead of cloning it with to_vec()
+        self.state.inner().inner().lines().to_vec()
```
Hmm this is perhaps an unfortunate consequence of the way I set up the traits, doesn't look like theres an easy way to do this

eadf replied:
So.. keep the comment as a future reminder?

eadf replied:
I had an idea that the DataRenderer could return a ref iterator, but I could not figure out how to to get a lifetime spec into the `Output` type w/o breaking backwards compat.

dcchut replied:
Yeah keeping it is fine.

dcchut commented on dcc-lsystem/Cargo.toml:
```diff
@@ -9,15 +9,71 @@ edition = "2018"
-color_quant = "1.0"
-gif = "0.11"
-tempfile = "3.1"
+# Todo: is color_quant used for anything?
```
Looks like it isn't - if the build passes with it removed it's probably fine.

eadf replied:
ok, removed color_quant

dcchut reviewed (commented):
Looks good, just a few things to consider.

eadf committed:
Removed the feature "data_renderer", the data renderer is now always included.
Simplified the construction of DataRendererOptions, if a builder pattern is required in the future it is easy to add.

eadf committed:
Removed the unused color_quant dependency.
Re-arranged so all optional dependencies are grouped together.

eadf commented:
Just noticed that the upper left side of the sierpinski_triangle example image looks weird. 
Is that something I did (by updating the dependencies maybe)? 

Edit: Left side of sierpinski_arrowhead is slightly shifted to the right too.
Edit2: Tested with `origin/master` and I got the same result, so that's a separate issue for later

eadf committed:
Moved all of the (optional) image renderer code into a separate module.

dcchut commented:
Nice one.  Thanks for your contribution!
